File: typescript/src/application/usecase/market_price/updateMarketPriceUsecase.ts
import Decimal from "decimal.js";
import { MarketPriceRepository } from "../../repository/marketPriceRepository.js";
import { StockSymbol } from "../../../domain/stockSymbol.js";

export interface UpdateMarketPriceItemInput {
  symbol: StockSymbol;
  marketPrice: Decimal;
}

export interface UpdateMarketPriceUsecaseInput {
  items: UpdateMarketPriceItemInput[];
}

export class UpdateMarketPriceUsecase {
  constructor(private readonly marketPriceRepository: MarketPriceRepository) {}

  async run(input: UpdateMarketPriceUsecaseInput): Promise<void> {
    const prices = new Map<StockSymbol, Decimal>(
      input.items.map((i) => [i.symbol, i.marketPrice]),
    );
    await this.marketPriceRepository.update(prices);
  }
}

File: golang/internal/application/usecase/market_price/update_market_price_usecase.go
package marketprice

import (
	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

type UpdateMarketPriceItemInput struct {
	Symbol      domain.StockSymbol
	MarketPrice decimal.Decimal
}

type UpdateMarketPriceUsecaseInput struct {
	Items []UpdateMarketPriceItemInput
}

type UpdateMarketPriceUsecase struct {
	marketPriceRepo repository.MarketPriceRepository
}

func NewUpdateMarketPriceUsecase(marketPriceRepo repository.MarketPriceRepository) *UpdateMarketPriceUsecase {
	return &UpdateMarketPriceUsecase{marketPriceRepo: marketPriceRepo}
}

func (u *UpdateMarketPriceUsecase) Run(input UpdateMarketPriceUsecaseInput) error {
	prices := make(map[domain.StockSymbol]decimal.Decimal, len(input.Items))
	for _, item := range input.Items {
		prices[item.Symbol] = item.MarketPrice
	}
	return u.marketPriceRepo.Update(prices)
}
